{
 "metadata": {
  "name": "",
  "signature": "sha256:c58a1c820df79ef981bab98b5136016330d47e09937432a3e1908cdb1c08055c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Overview of examples that we will be working through in coming Python lessons"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Example 1: Working with DNA strings**\n",
      "\n",
      "Suppose we are given a DNA string. We may like to know the following:\n",
      "\n",
      "1. What is its complement?\n",
      "2. What is its reverse complemment?\n",
      "3. What would the sequence of the RNA transcript be?\n",
      "4. What is the GC-content?\n",
      "\n",
      "```python\n",
      "dna = 'GATTACA'\n",
      "\n",
      "# Question 1\n",
      "import string\n",
      "comp_table = string.maketrans('ATCG', 'TAGC')\n",
      "comp_dna = dna.translate(comp_table)\n",
      "print comp_dna\n",
      "\n",
      "# Question 2\n",
      "rev_comp_dna = comp_dna[::-1]\n",
      "print rev_comp_dna\n",
      "\n",
      "# Question 3\n",
      "rna = dna.replace('T', 'U')\n",
      "print rna\n",
      "\n",
      "# Question 4\n",
      "num_gc = dna.count('G') + dna.count('C')\n",
      "pct_gc = 100*(num_gc/float(len(dna)))\n",
      "print pct_gc\n",
      "```"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Example 2: Exact M.W. of ssDNA (e.g., Oligonucleotides)**\n",
      "\n",
      "Suppose we are give a ssDNA sequence 'GATTACA'. We want to write a function calculate the molecular weight of any such sequence using the following formula: \n",
      "\n",
      "M.W. = (An x 313.2) + (Tn x 304.2) + (Cn x 289.2) + (Gn x 329.2) + 79.0\u00aa\n",
      "An, Tn, Cn, and Gn are the number of each respective nucleotide within the polynucleotide.\n",
      "\u00aaAddition of \"79.0\" to the M.W. takes into account the 5' monophosphate left by most restriction enzymes. No phosphate is present at the 5' end of strands made by primer extension.\n",
      "\n",
      "<http://www.lifetechnologies.com/us/en/home/references/ambion-tech-support/rna-tools-and-calculators/dna-and-rna-molecular-weights-and-conversions.html>\n",
      "\n",
      "```python\n",
      "dna_mw = {\n",
      "    'A' : 313.2,\n",
      "    'T' : 304.2, \n",
      "    'C' : 289.2,\n",
      "    'G' : 329.2\n",
      "    }\n",
      "\n",
      "def mw(seq, mw_dict, offset=79.04):\n",
      "    \"\"\"Return molecular weight of the ssDNA string seq.\"\"\"\n",
      "    wt = sum(mw_dict[nuc] for nuc in seq) + offset\n",
      "    return wt\n",
      "\n",
      "dna = 'GATTACA'\n",
      "print mw(dna, dna_mw)\n",
      "```"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Example 3: Find the frequncy of a specific DNA sequence in a collection of many DNA sequences (e.g. from fragmentation)**\n",
      "\n",
      "Suppose we have sequences from 1 million short DNA fragments. We want to answer some of the following questions:\n",
      "\n",
      "1. What is the frequency of the fragment 'GATTACA'?\n",
      "2. What is the distribution of fragment lengths?\n",
      "3. What are the 3 most common fragments?\n",
      "\n",
      "```python\n",
      "\n",
      "# Question 1 - manual solution\n",
      "def count_fragments(frags):\n",
      "    \"\"\"Return a dictionary where the key is the fragment sequence and values is number of times it is found.\"\"\"\n",
      "    counts = {}\n",
      "    for frag in frags:\n",
      "        counts[frag] = counts.get(frag, 0) + 1\n",
      "    return counts\n",
      "\n",
      "counts = count_fragments(frags)\n",
      "print counts['GATTACA']\n",
      "\n",
      "# Question 2 - using collections.Counter\n",
      "frag_lens = [len(frag) for frag in frags]\n",
      "from collections import Counter\n",
      "print Counter(frag_lens)\n",
      "\n",
      "# Question 3 - using collections.Counter\n",
      "from collections import Counter\n",
      "print Counter(frags).most_common(3)\n",
      "```"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Example 4: Finding ORFs and transcription products\n",
      "\n",
      "Suppose we are given a DNA sequence. \n",
      "\n",
      "1. Can we find an open reading frame within the sequence?\n",
      "2. What would the amino acid sequence be of the protein translated from this ORF?\n",
      "\n",
      "```python\n",
      "# Question 1\n",
      "def find_orf(seq):\n",
      "    \"\"\"Returns list of codons in ORF or the empty list if an ORF is not found.\"\"\"\n",
      "      \n",
      "    start_codon = 'AUG'\n",
      "    stop_codons = ['TAG', 'TAA', 'TGA']\n",
      "\n",
      "    stop_idx = 0\n",
      "    start_idx = seq.find(start_codon)\n",
      "    \n",
      "    # if start codon is not found, there is no ORF so return empty list\n",
      "    if start_idx == -1:\n",
      "        return []\n",
      "    \n",
      "    # possile codons are the 3-mers beginning with the start codon\n",
      "    possible_codons = [seq[i:(i+3)] for i in range(start_idx, len(seq), 3)]\n",
      "\n",
      "    # find stop codon if it is among possibe codons and return (possible empty) list of codons\n",
      "    stop = 0\n",
      "    for stop_codon in stop_codons:\n",
      "        if stop_codon in possible_codons:\n",
      "            stop = possible_codons.index(stop_codon)\n",
      "\n",
      "    return possible_codons[:stop]\n",
      "\n",
      "# DNA string is left flank + start + bunch of codons + stop + right flank\n",
      "dna = 'GT' + 'AUG' + 'CCACGTTCCCATATCAGTGAT' + 'TGA' + 'TA'\n",
      "orf = find_orf(dna)\n",
      "\n",
      "# Question 2\n",
      "from seq_utils import codon_map\n",
      "\n",
      "protein = [codon_map[codon] for codon in orf]\n",
      "print protein\n",
      "```"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Utilty modules"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file seq_utils.py\n",
      "\n",
      "codon_map = {\n",
      " 'AAA': 'K',\n",
      " 'AAC': 'N',\n",
      " 'AAG': 'K',\n",
      " 'AAT': 'N',\n",
      " 'ACA': 'T',\n",
      " 'ACC': 'T',\n",
      " 'ACG': 'T',\n",
      " 'ACT': 'T',\n",
      " 'AGA': 'R',\n",
      " 'AGC': 'S',\n",
      " 'AGG': 'R',\n",
      " 'AGT': 'S',\n",
      " 'ATA': 'I',\n",
      " 'ATC': 'I',\n",
      " 'ATG': 'M',\n",
      " 'ATT': 'I',\n",
      " 'CAA': 'Q',\n",
      " 'CAC': 'H',\n",
      " 'CAG': 'Q',\n",
      " 'CAT': 'H',\n",
      " 'CCA': 'P',\n",
      " 'CCC': 'P',\n",
      " 'CCG': 'P',\n",
      " 'CCT': 'P',\n",
      " 'CGA': 'R',\n",
      " 'CGC': 'R',\n",
      " 'CGG': 'R',\n",
      " 'CGT': 'R',\n",
      " 'CTA': 'L',\n",
      " 'CTC': 'L',\n",
      " 'CTG': 'L',\n",
      " 'CTT': 'L',\n",
      " 'GAA': 'E',\n",
      " 'GAC': 'D',\n",
      " 'GAG': 'E',\n",
      " 'GAT': 'D',\n",
      " 'GCA': 'A',\n",
      " 'GCC': 'A',\n",
      " 'GCG': 'A',\n",
      " 'GCT': 'A',\n",
      " 'GGA': 'G',\n",
      " 'GGC': 'G',\n",
      " 'GGG': 'G',\n",
      " 'GGT': 'G',\n",
      " 'GTA': 'V',\n",
      " 'GTC': 'V',\n",
      " 'GTG': 'V',\n",
      " 'GTT': 'V',\n",
      " 'TAA': '*',\n",
      " 'TAC': 'Y',\n",
      " 'TAG': '*',\n",
      " 'TAT': 'Y',\n",
      " 'TCA': 'S',\n",
      " 'TCC': 'S',\n",
      " 'TCG': 'S',\n",
      " 'TCT': 'S',\n",
      " 'TGA': '*',\n",
      " 'TGC': 'C',\n",
      " 'TGG': 'W',\n",
      " 'TGT': 'C',\n",
      " 'TTA': 'L',\n",
      " 'TTC': 'F',\n",
      " 'TTG': 'L',\n",
      " 'TTT': 'F'}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Writing seq_utils.py\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}